File: golang/pub-sub/poster.go
package main

import (
	"encoding/json"
	"flag"
	"log"
	"net/http"

	"github.com/nats-io/nats.go"
)

// Define the structure for your JSON data
type Data struct {
	Name  string `json:"name"`
	Age   int    `json:"age"`
	Email string `json:"email"`
}

func main() {
	port := flag.Int("port", 5000, "Port on which to listen")
	natsServer := flag.String("nats-server", "nats://localhost:4222", "NATS server address")
	natsSubject := flag.String("nats-subject", "hello.world", "NATS subject to publish to")
	flag.Parse()

	http.HandleFunc("/v1/api", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			http.Error(w, "Only POST method is allowed", http.StatusMethodNotAllowed)
			return
		}

		var data Data
		err := json.NewDecoder(r.Body).Decode(&data)
		if err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}

		// Here, you would add additional validation for the Data struct if needed

		nc, err := nats.Connect(*natsServer)
		if err != nil {
			log.Printf("Failed to connect to NATS server: %v", err)
			http.Error(w, "Failed to connect to NATS server", http.StatusInternalServerError)
			return
		}
		defer nc.Close()

		dataBytes, err := json.Marshal(data)
		if err != nil {
			http.Error(w, "Error encoding JSON", http.StatusInternalServerError)
			return
		}

		err = nc.Publish(*natsSubject, dataBytes)
		if err != nil {
			log.Printf("Failed to publish message: %v", err)
			http.Error(w, "Failed to publish message", http.StatusInternalServerError)
			return
		}

		log.Printf("Published message to %s: %+v", *natsSubject, data)
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(map[string]string{"message": "Data received and published successfully"})
	})

	log.Printf("Server starting on port %d...", *port)
	log.Fatal(http.ListenAndServe(":5000", nil))
}

File: python/pub-sub/poster.py
from flask import Flask, request, jsonify
from jsonschema import validate, ValidationError
import asyncio
import json
import logging
import nats
from nats.errors import TimeoutError, ConnectionClosedError, NoServersError
import argparse

# Setup argparse for command-line arguments
parser = argparse.ArgumentParser(description='Run a Flask application with configurable NATS integration.')
parser.add_argument('--port', type=int, default=5000, help='Port on which to listen')
parser.add_argument('--nats-server', type=str, default='nats://localhost:4222', help='NATS server address')
parser.add_argument('--nats-subject', type=str, default='hello.world', help='NATS subject to publish to')
args = parser.parse_args()

app = Flask(__name__)

# JSON Schema for validation
schema = {
    "type": "object",
    "properties": {
        "name": {"type": "string"},
        "age": {"type": "integer"},
        "email": {"type": "string", "format": "email"}
    },
    "required": ["name", "age", "email"]
}

# Configure logging
logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__name__)

@app.route('/v1/api', methods=['POST'])
async def handle_request():
    nc = await nats.connect(args.nats_server)  # Connect to NATS server
    try:
        data = request.get_json()
        validate(instance=data, schema=schema)  # Validate JSON against schema
        await nc.publish(args.nats_subject, json.dumps(data).encode('utf-8'))  # Publish message to "hello.world" subject
        logger.debug("Published message to %s: %s", args.nats_subject, data)
        return jsonify({"message": "Data received and published successfully"}), 200
    except ValidationError as e:
        return jsonify({"error": str(e)}), 400  # Bad request if JSON schema validation fails
    except NoServersError as e:
        logger.error("No NATS servers available for connection: %s", str(e))
        return jsonify({"error": "Service unavailable, please try again later."}), 503  # Service Unavailable
    except ConnectionClosedError as e:
        logger.error("Connection to NATS was closed unexpectedly: %s", str(e))
        return jsonify({"error": "Connection lost, please try again."}), 503  # Service Unavailable
    except TimeoutError as e:
        logger.error("Error publishing message: %s", str(e))
        return jsonify({"error": str(e)}), 500  # Internal server error if NATS issues occur

if __name__ == '__main__':
    loop = asyncio.get_event_loop()
    app.run(debug=True, port=args.port)
